### ISSUE TITLE
A better approach to test Koreader for Android issues

### ISSUE THREAD

**Issue by chrox** [koreader/koreader#2113]
I found a Android auto testing platform https://mqc.aliyun.com that could help us finally resolve the crashes and ANR problems on Android. The main difficulty we have to investigate into this, I think, is that we don't have enough Android devices to reproduce those problems. And with the help of mqc.aliyun.com we can upload the APK of Koreader to that platform and it will be tested on 30 devices in one task and generated logcats are available for further investigation. The latest nightly build got a  55% pass rate (or a little higher considering there are two Intel Atom devices that have different CPU architecture with Koreader app). And the auto tests also helped me find two bugs as fixed in #2112.

What do you think of this? And if we go to 95% pass rate could we close #2027, #1904, #1692, #673 and #643?

Below is a screenshot of the testing platform:
![download](https://cloud.githubusercontent.com/assets/751535/16388581/5d27c042-3ccc-11e6-834e-ecefc0ccbb96.png)

Sorry that this platform is Chinese language only.

**Comment by Frenzie** [comment]
Looks very nice. What kind of tests does it do, exactly?

**Comment by chrox** [comment]
It's a kind of monkey testing, random inputs are generated by some kind of a program for 5 minutes util the application crashes. Since our primary problem on Android platform is that the app just don't run, the monkey testing is a perfect approach.

**Comment by houqp** [comment]
This is great tool! I am not able to find other comparable free offering that's in English. How do they charge for the service?

**Comment by chrox** [comment]
It's free for the first three testing tasks everyday. And each free testing task will be ran on 30 randomly selected devices. They will charge for additional services like manually select on which devices the test will be ran or including more devices in each testing task. But for us, the free service is good enough.

**Comment by pazos** [comment]
I found https://monkop.com, which does something similar (but have less devices available). 

I ran a test on friday nightly and the monkey test did not found any ANR but just one error.

![Screenshot_20190311_202940](https://user-images.githubusercontent.com/975883/54152020-77ca8b80-443c-11e9-9571-15632869261f.png)

**Comment by Frenzie** [comment]
If it's free for open source projects (it only says something about free for beta purposes?) it could probably be added as a deploy step on GitLab. (Or `after_success` like on Travis if GitLab has that.)

**Comment by NiLuJe** [comment]
Those are some mightily old Samsung devices, though ;).

**Comment by pazos** [comment]
> If it's free for open source projects (it only says something about free for beta purposes?) it could probably be added as a deploy step on GitLab. (Or after_success like on Travis if GitLab has that.)

I think they're not providing a free plan besides the beta testing.

> Those are some mightily old Samsung devices, though ;).

Yup, that's true!

**Comment by Frenzie** [comment]
Monkop sent me an e-mail that they're changing their name to https://www.apptim.com/

**Comment by pazos** [comment]
For random testing (after doing changes in the luajit-launcher) I'm using https://developer.android.com/studio/test/monkey, which is nice but random. I didn't find a seed that is able to reproduce the only ANR that I can find:

1. connect to looooow speed wifi
2. look for updates
3. generate a bunch (about 100 in less than a second) of touch events
4. profit.

**Comment by Frenzie** [comment]
I don't know how you'd go about generating more than a handful of touch events in a second (swiping?) but isn't the network thing just setting the network to something older (like GPRS or even GSM) in combination with network strength poor?

![Screenshot_2019-08-02_17-24-36](https://user-images.githubusercontent.com/202757/62380915-9267e300-b54a-11e9-8229-e8076b652a2b.png)

Admittedly I haven't tested that. If it doesn't work you could just run the Android device through a proxy on your desktop. 

http://www.dallaway.com/sloppy/ (older, might be what I used a decade ago)
https://github.com/tylertreat/Comcast
https://github.com/bcoe/crapify

**Comment by pazos** [comment]
Maybe not 100 in less a second, but a few while the "checking updates" Infomessage is shown

I'm using
```
adb shell
while true; do input touchpad swipe 100 100 200 200 1; done
```

But now that I check the easiest way to reproduce that is when downloading a dictionary. See what happens while trying to download a medium size dict (and a lot of touchpad swipe events are generated): 
![device-2019-08-02-180318](https://user-images.githubusercontent.com/975883/62383356-d90c0c00-b54f-11e9-887f-af624d1d7374.png)

**Comment by Frenzie** [comment]
I see. That's expected although hardly ideal on any platform, but Android expects a response. One might call it a big fundamental issue, but if there's an easy Android-specific workaround that reduces that I'd say why not. ;-)

**Comment by pazos** [comment]
> I see. That's expected although hardly ideal on any platform, but Android expects a response. One might call it a big fundamental issue, but if there's an easy Android-specific workaround that reduces that I'd say why not. ;-)

Sadly we lost the main benefit of the app_native_glue (the hability to poll and react to events without blocking the ui thread) by running the looper via luajit FFI. The problem is that, when luajit is busy elsewhere, we cannot drain events from the InputQueue and finally the ActivityManager triggers the ANR.  This shouldn't happen if we use the Alooper inside the native library (aka android_main.c) but I have no idea how process and pass the events to the frontend. This seems a job for qualified people like @hwhw, @chrox or @houqp.

Anyways the system works flawless most of the time and just a few conditions (reflowing, opening a complex epub for the first time, network stuff...) block luajit enough to trigger the ANR. In these cases I cannot see a workaround coded in lua because it will be blocked too. Maybe I'm wrong. There is hope 😄

**Comment by arooni** [comment]
Following in the hope this issue resolves issues like #5665 and #5020

**Comment by pazos** [comment]
I'm going to close this issue. The original topic: test KOReader on a bunch of android devices is probably out of date. The few ANRs monkey tests are triggering are related to https://github.com/koreader/koreader/issues/2113#issuecomment-517858265 and can be workarounded with https://github.com/koreader/koreader/pull/6266.
